drivers/adc/ad7091r8/ad7091r8.c

```diff
@@ -0,0 +1,609 @@
+		value = NO_OS_BIT(4) | NO_OS_BIT(5);
+		break;
+	default:
+		return -EINVAL;
+	}
+	if (is_cmos)
+		value |= NO_OS_BIT(6);
+
+	return ad7091r8_spi_write_mask(dev, AD7091R8_REG_CONF,
+				       REG_CONF_GPO0_MODE_MASK, value);
+}
+
+/**
+ * Set high limit, low limit, hysteresis.
+ * Device accepts 9 bits provided by the user and sets them as the MSBs.
+ * The 3 LSBs of the internal 12-bit registers are set either to 000 or 111.
+ * Round user input according to each case.
+ * @param dev - The device structure.
+ * @param limit - Limit.
+ * @param channel - Channel.
+ * @param value - Value.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_set_limit(struct ad7091r8_dev *dev,
+		       enum ad7091r8_limit limit,
+		       uint8_t channel,
+		       uint16_t value)
+{
+	uint16_t reg;
+
+	if (!dev)
+		return -EINVAL;
+
+	switch (limit) {
+	case AD7091R8_LOW_LIMIT:
+		reg = AD7091R8_REG_CH_LOW_LIMIT(channel);
+		/* Approximate reg value to requested limit value */
+		if (value & 0x04)
+			value += 0x08;
+		break;
+	case AD7091R8_HIGH_LIMIT:
+		reg = AD7091R8_REG_CH_HIGH_LIMIT(channel);
+		/* Approximate reg value to requested limit value */
+		if (value & 0x04)
+			value -= 0x08;
+		break;
+	case AD7091R8_HYSTERESIS:
+		reg = AD7091R8_REG_CH_HYSTERESIS(channel);
+		/* Approximate reg value to requested hysteresis value */
+		if (value & 0x04)
+			value += 0x08;
+		break;
+	default:
+		return -EINVAL;
+	}
+	value = value >> 3;
+
+	return ad7091r8_spi_reg_write(dev, reg,
+				      no_os_field_get(AD7091R8_CONV_MASK, value));
+}
+
+/**
+ * Get alert.
+ * @param dev - The device structure.
+ * @param channel - Channel.
+ * @param alert - Alert type.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_get_alert(struct ad7091r8_dev *dev, uint8_t channel,
+		       enum ad7091r8_alert_type *alert)
+{
+	int ret;
+	uint16_t data;
+
+	if (!dev || !alert)
+		return -EINVAL;
+
+	if (channel >= AD7091R_NUM_CHANNELS(dev->device_id))
+		return -EINVAL;
+
+	ret = ad7091r8_spi_reg_read(dev, AD7091R8_REG_ALERT, &data);
+	if (ret)
+		return ret;
+
+	*alert = REG_ALERT_MASK(data, channel);
+
+	return 0;
+}
+
+/**
+ * Get high limit, low limit, hysteresis.
+ * The 3 LSBs of the internal 12-bit registers are set either to 000 or 111.
+ * Adjust limit data to reflect actual limit in use by the device.
+ * @param dev - The device structure.
+ * @param limit - Limit.
+ * @param channel - Channel.
+ * @param value - Value.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_get_limit(struct ad7091r8_dev *dev,
+		       enum ad7091r8_limit limit,
+		       uint8_t channel,
+		       uint16_t *value)
+{
+	int ret;
+	uint16_t reg, data;
+
+	if (!dev || !value)
+		return -EINVAL;
+
+	switch (limit) {
+	case AD7091R8_LOW_LIMIT:
+		reg = AD7091R8_REG_CH_LOW_LIMIT(channel);
+		break;
+	case AD7091R8_HIGH_LIMIT:
+		reg = AD7091R8_REG_CH_HIGH_LIMIT(channel);
+		*value = 0x07;
```

Comment:
Please check all the code for places where magic numbers can be replaced by defines.